### PR TITLE
make internal-reflection indy compatible

### PR DESCRIPTION
**Pull request by SylvainJuge**
Part of #11457 

Migrates `internal-reflection` instrumentation module to be indy compatible.

Also includes two bugfixes that prevented injecting proxy classes in bootstrap classloader.

-----

Challenge(s) ~~to solve~~ solved with 6747406431b4a725718c0223a9b1f9aefa6393ba and cae6644

For hotspot & openJ9 compatibility, extra invocation of `ReflectionHelper` is done directly with ASM within `java.lang.Class#getInterfaces`

From what I understand, the original values are from native code (or equivalent for OpenJ9), so it should be simpler to just inject the helper class in the classloader, however:

Using `ExperimentalInstrumentationModule#injectedClassNames` does not work and triggers `Nested instrumented invokedynamic instruction linkage detected` warning.

Using `ExperimentalInstrumentationModule#injectClasses` does not work either and we get an exception that might indicate it does not work for the bootstrap classloader:

```
java.lang.IllegalArgumentException: io.opentelemetry.javaagent.instrumentation.internal.reflection.ReflectionInstrumentationModule@191ae03f has not been initialized for class loader <bootstrap> yet
        at io.opentelemetry.javaagent.tooling.instrumentation.indy.IndyModuleRegistry.getInstrumentationClassLoader(IndyModuleRegistry.java:52)
        at io.opentelemetry.javaagent.tooling.instrumentation.indy.ClassInjectorImpl$ProxyBuilder.lambda$inject$0(ClassInjectorImpl.java:62)
        at io.opentelemetry.javaagent.tooling.instrumentation.indy.ClassInjectorImpl.lambda$getClassesToInject$0(ClassInjectorImpl.java:38)

```

For this second issue, I think the problem is in `io.opentelemetry.javaagent.tooling.instrumentation.indy.IndyModuleRegistry#getInstrumentationClassLoader(io.opentelemetry.javaagent.extension.instrumentation.InstrumentationModule, java.lang.ClassLoader)` where a call is made to a `ClassLoaderValue` with the wrong key.
The argument for helper classes is `HelperInjector#BOOTSTRAP_CLASSLOADER_PLACEHOLDER` whereas `ClassLoaderValue` is using `null` to indicate bootstrap classloader.

Update: this was solved with cae6644

After fixing this issue, we get back to the nested invokedynamic issue we had initially :-(.

### CODE REVIEW

File: instrumentation/internal/internal-reflection/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/internal/reflection/ReflectionInstrumentationModule.java [hunk 1]
```diff
@@ -26,12 +30,20 @@ public boolean defaultEnabled(ConfigProperties config) {
...
-    return false;
+  public List<TypeInstrumentation> typeInstrumentations() {
+    return asList(new ClassInstrumentation(), new ReflectionInstrumentation());
   }
 
   @Override
-  public List<TypeInstrumentation> typeInstrumentations() {
-    return asList(new ClassInstrumentation(), new ReflectionInstrumentation());
+  public String getModuleGroup() {
+    return "internal-reflection";
```
Comment (JonasKunz) [inline instrumentation/internal/internal-reflection/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/internal/reflection/ReflectionInstrumentationModule.java:39]: Is the module group here necessary? It seems to be the only module in that group, in that case there is no need to override this method

Comment (SylvainJuge) [reply instrumentation/internal/internal-reflection/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/internal/reflection/ReflectionInstrumentationModule.java:39]: Good point, it's not required here. I'll wait for the tests to complete in CI and remove it once they pass.

File: muzzle/src/main/java/io/opentelemetry/javaagent/tooling/HelperInjector.java [hunk 2]
```diff
@@ -172,7 +172,8 @@ public DynamicType.Builder<?> transform(
     injectedClassLoaders.computeIfAbsent(
         maskNullClassLoader(classLoader),
         cl -> {
-          List<HelperClassDefinition> helpers = helperClassesGenerator.apply(cl);
+          List<HelperClassDefinition> helpers =
+              helperClassesGenerator.apply(isBootClassLoader(cl) ? null : cl);
```
Comment (laurit) [inline muzzle/src/main/java/io/opentelemetry/javaagent/tooling/HelperInjector.java:176]: perhaps add a method `unmaskNullClassLoader` to make it clear that this here reverses the previous call to `maskNullClassLoader`

Comment (SylvainJuge) [reply muzzle/src/main/java/io/opentelemetry/javaagent/tooling/HelperInjector.java]: Good idea, I've added it in https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/12136/commits/e5bfd05b8e63b29603853e4ce73e1dde6b1bcb0f

**Review by laurit — APPROVED** [review]

File: instrumentation/internal/internal-reflection/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/internal/reflection/ReflectionInstrumentationModule.java [hunk 3]
```diff
@@ -26,12 +30,15 @@ public boolean defaultEnabled(ConfigProperties config) {
...
+    return asList(new ClassInstrumentation(), new ReflectionInstrumentation());
   }
 
   @Override
-  public List<TypeInstrumentation> typeInstrumentations() {
-    return asList(new ClassInstrumentation(), new ReflectionInstrumentation());
+  public void injectClasses(ClassInjector injector) {
+    injector
+        .proxyBuilder(
+            "io.opentelemetry.javaagent.instrumentation.internal.reflection.ReflectionHelper")
```
Comment (trask) [inline instrumentation/internal/internal-reflection/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/internal/reflection/ReflectionInstrumentationModule.java:41]: can you add a comment here, similar to https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/814c3ea9fadd799bc12c4d268a3ea23ca841218d/instrumentation/graphql-java/graphql-java-20.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/graphql/v20_0/GraphqlInstrumentationModule.java#L42-L44?

**Review by trask — APPROVED** [review]

File: instrumentation/internal/internal-reflection/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/internal/reflection/ReflectionInstrumentationModule.java (hunk 3 above)
Comment (SylvainJuge) [reply instrumentation/internal/internal-reflection/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/internal/reflection/ReflectionInstrumentationModule.java:44]: Fixed in https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/12136/commits/9bcb526bb1e4ba767b853b750fd788ca5e979afa

**Pull request merged**